daemon/cmd/daemon.go

```diff
@@ -208,6 +213,28 @@ type Daemon struct {
 	bigTCPConfig bigtcp.Configuration
 }
 
+func (d *Daemon) initDNSProxyContext(size int) {
+	d.dnsProxyContext = dnsProxyContext{
+		responseMutexes: make([]*lock.Mutex, size),
+		modulus:         big.NewInt(int64(size)),
+	}
+	for i := range d.dnsProxyContext.responseMutexes {
+		d.dnsProxyContext.responseMutexes[i] = new(lock.Mutex)
+	}
+}
+
+// dnsProxyContext is a meta struct containing fields relevant to the DNS proxy
+// of the daemon, for organizational purposes.
+type dnsProxyContext struct {
+	// responseMutexes is used to serialized the critical path of
+	// notifyOnDNSMsg() to ensure that identity allocation and ipcache
+	// insertion happen atomically between parallel DNS request handling.
+	responseMutexes []*lock.Mutex
+	// modulus is used when computing the hash of the DNS response IPs in order
+	// to map them to the mutexes inside responseMutexes.
+	modulus *big.Int
```

Comment:
Why does the modulus need to be `big.Int`?

Reply:
Context: 32 bits should already be plenty. We might even get away with 16 bits (65K) locks, at which point there's a 0.5MB memory cost. Even for handling policy for kube-dns on the order of 10krps this should _probably_ be reasonable, though we may want to bump that a bit beyond 16bits at some point for such specialized workloads.

Reply:
Unfortunately, the `big.Int` API only allows you to take the `Mod()` with a `big.Int` parameter. So instead of constructing the modulus via `big.NewInt()` on every call to `getMutexesForResponseIPs()`, I opted to just store it like this upfront.

Reply:
Why do we need big.Int at all? For modulus operations?

Reply:
Ah, so we need `big.Int` in order to represent IPv6 as an integer. We represent IPs as integers because `net.IP` types are not hashable, so they cannot be map keys, therefore we convert them into integers (via byte arrays as that's the backing structure behind `net.IP`). Obviously an IPv4 address is only 32 bits, but IPv6 is 128 bits. Go doesn't have a native 128 bit, see https://github.com/golang/go/issues/9455. If that proposal ever lands, then we can move away from `big.Int`.

Reply:
I see. The "`net.IP` types are not hashable" aspect could also be addressed by the newer `netip.Addr` types but then I guess we may not be able to backport this fix to all branches. Personally I'd probably prefer this for a longer term solution, but if we're trying to avoid bumping the golang library dependency minor version in older releases then `big.Int` seems like a reasonable intermediate 👍

Reply:
Yep that was the reason to use `big.Int` for now. I agree on moving towards `netip.Addr` when possible.